### PR TITLE
Add a signal before building patches for visualization output

### PR DESCRIPTION
**Pull request by anne-glerum**
Add a signal that can modify data_out during the creation of visualization output. It can be used to for example select only part of the domain to be outputted. 

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: tests/pre_data_out_build_patches.cc [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include <aspect/simulator_access.h>
+
+#include <iostream>
+
+namespace aspect
+{
+  template <int dim>
+  void pre_data_out_build_patches (DataOut<dim> &)
+  {
+    std::cout << "\npre_data_out_build_patches:\n";
```
[the remark this answers is not in the document]
Comment (anne-glerum) [reply tests/pre_data_out_build_patches.cc:31]: It is triggered. The test output I committed is still from the test I copied from. But you can see that "npre_data_out_build_patches:" is printed by the tester. See attached figure. I'll update with the results from the tester.
Screenshot 2024-06-05 at 16 58 25

File: include/aspect/simulator_signals.h [hunk 2]
```diff
@@ -287,6 +288,15 @@ namespace aspect
...
     set_assemblers;
+
+    /**
+    * A signal that is called before the build_patches() function is called during
+    * the creation of the visualization output. This signal
+    * allows for registering functions that take a DataOut object and can for example
+    * be used to select only certain cells of the mesh to be built into patches through
+    * calling the DataOut member function set_cell_selection().
+    */
+    boost::signals2::signal<void (DataOut<dim> &)>  pre_data_out_build_patches;
```
Comment (bangerth) [inline include/aspect/simulator_signals.h:299]: This is one way. I assume you want to use this to let the called function attach itself to the DataOut? Why not have a signal that returns a function object selecting individual cells that can then directly be attached to DoFHandler?

Comment (anne-glerum) [reply include/aspect/simulator_signals.h:299]: As just discussed, we tried to keep the signal as general as possible as other things could be done with DataOut.

**Review by bangerth — APPROVED** [review]

**Review by tjhei — APPROVED** [review]

**Review by bangerth — CHANGES_REQUESTED** [review]
I forgot: Would you mind reducing the number of commits?

**Review by bangerth — APPROVED** [review]

**Pull request merged**